File: week1/week_test/linkTest/std_io.cpp
#include <iostream>
#include <vector>
using namespace std;
// 代码实现开始
const int N = 1005;

struct Node
{
    int pre, next;
    Node(int a, int b)
    {
        pre = a;
        next = b;
    }
};
vector<Node> my_list;
// 初始化函数，在操作开始前会调用一次
void init(int n)
{
    my_list.assign(n + 1, Node(0, 0));
    // for (int i = 0; i < n; i++)
    // {
    //     my_list.push_back(Node(0, 0));
    // }
}

// x 是编号，范围为 1 到 n
// 若成功操作，返回 true，否则返回 false
bool split_succ(int x)
{
    if (my_list[x].next)
    {
        int t = my_list[x].next;
        my_list[t].pre = 0;
        my_list[x].next = 0;
        return true;
    }
    return false;
}

// x 是编号，范围为 1 到 n
// 若成功操作，返回 true，否则返回 false
bool split_prev(int x)
{
    if (my_list[x].pre)
    {
        int t = my_list[x].pre;
        my_list[x].pre = 0;
        my_list[t].next = 0;
        return true;
    }
    return false;
}

// x, y 是编号，范围为 1 到 n
// 若成功操作，返回 true，否则返回 false
bool link(int x, int y)
{
    if(!my_list[x].next && !my_list[y].pre){
        my_list[x].next = y;
        my_list[y].pre = x;
        return true;
    }
    return false;
}

// x 是编号，范围为 1 到 n
// 返回遍历得到的序列
std::vector<int> visit_succ(int x)
{
    std::vector<int> ans;
    ans.push_back(x);
    int af = my_list[x].next;
    while (af && af!=x){
        ans.push_back(af);
        af = my_list[af].next;
    }
        return ans;
}

// x 是编号，范围为 1 到 n
// 返回遍历得到的序列
std::vector<int> visit_prev(int x)
{
    std::vector<int> ans;
    ans.push_back(x);
    int bf = my_list[x].pre;
    while (bf && bf != x)
    {
        ans.push_back(bf);
        bf = my_list[bf].pre;
    }
    return ans;
}

// 代码实现结束

int main()
{
    std::ios::sync_with_stdio(false);
    int n, m, x, y;
    std::string op;
    std::cin >> n >> m;
    init(n);
    for (int i = 0; i < m; ++i)
    {
        std::cin >> op >> x;
        if (op == "split_succ")
        {
            std::cout << (split_succ(x) ? "yes" : "no") << '\n';
        }
        else if (op == "split_prev")
        {
            std::cout << (split_prev(x) ? "yes" : "no") << '\n';
        }
        else if (op == "link")
        {
            std::cin >> y;
            std::cout << (link(x, y) ? "yes" : "no") << '\n';
        }
        else if (op == "visit_succ")
        {
            std::vector<int> ans = visit_succ(x);
            for (int i = 0; i < ans.size(); i++)
            {
                std::cout << ans[i] << " \n"[i + 1 == ans.size()];
            }
        }
        else
        {
            std::vector<int> ans = visit_prev(x);
            for (int i = 0; i < ans.size(); i++)
            {
                std::cout << ans[i] << " \n"[i + 1 == ans.size()];
            }
        }
    }
    return 0;
}

File: week1/week_test/linkTest/std_io.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) {
        InputStream inputStream = System.in;
        OutputStream outputStream = System.out;
        InputReader in = new InputReader(inputStream);
        PrintWriter out = new PrintWriter(outputStream);
        Task solver = new Task();
        solver.solve(in, out);
        out.close();
    }

    static class Task {

        // ================= 代码实现开始 =================

        // 初始化函数，在操作开始前会调用一次
        void init(int n) {
        }

        // x 是编号，范围为 1 到 n
        // 若成功操作，返回 true，否则返回 false
        boolean split_succ(int x) {
            return true;
        }

        // x 是编号，范围为 1 到 n
        // 若成功操作，返回 true，否则返回 false
        boolean split_prev(int x) {
            return true;
        }

        // x, y 是编号，范围为 1 到 n
        // 若成功操作，返回 true，否则返回 false
        boolean link(int x, int y) {
            return true;
        }

        // x 是编号，范围为 1 到 n
        // 返回遍历得到的序列
        List<Integer> visit_succ(int x) {
            return new ArrayList<>();
        }

        // x 是编号，范围为 1 到 n
        // 返回遍历得到的序列
        List<Integer> visit_prev(int x) {
            return new ArrayList<>();
        }

        // ================= 代码实现结束 =================

        void solve(InputReader in, PrintWriter out) {
            int n = in.nextInt();
            int m = in.nextInt();
            init(n);
            for (int i = 0; i < m; ++i) {
                String op = in.next();
                int x = in.nextInt();
                if (op.equals("split_succ")) {
                    out.println(split_succ(x) ? "yes" : "no");
                } else if (op.equals("split_prev")) {
                    out.println(split_prev(x) ? "yes" : "no");
                } else if (op.equals("link")) {
                    int y = in.nextInt();
                    out.println(link(x, y) ? "yes" : "no");
                } else if (op.equals("visit_succ")) {
                    List<Integer> ans = visit_succ(x);
                    for (Integer t : ans) {
                        out.print(t);
                        out.print(' ');
                    }
                    out.println();
                } else {
                    List<Integer> ans = visit_prev(x);
                    for (Integer t : ans) {
                        out.print(t);
                        out.print(' ');
                    }
                    out.println();
                }
            }
        }

    }

    static class InputReader {
        BufferedReader reader;
        StringTokenizer tokenizer;

        private InputReader(InputStream stream) {
            reader = new BufferedReader(new InputStreamReader(stream), 32768);
            tokenizer = null;
        }

        private String next() {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                try {
                    tokenizer = new StringTokenizer(reader.readLine());
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
            return tokenizer.nextToken();
        }

        private short nextShort() {
            return Short.parseShort(next());
        }

        private int nextInt() {
            return Integer.parseInt(next());
        }

        private long nextLong() {
            return Long.parseLong(next());
        }

        private float nextFloat() {
            return Float.parseFloat(next());
        }

        private double nextDouble() {
            return Double.parseDouble(next());
        }
    }

}